src/main.jsligo

```diff
@@ -0,0 +1,182 @@
+      let {owner,tokenId,amount_} = paramBurn;
+      [
+        FA2.Ledger.decrease_token_amount_for_user(ledger)(owner)(tokenId)(amount_),
+        Extension.decreaseSupply(supply, tokenId, amount_)
+      ]
+    };
+   let [ledger, supply] = List.fold_left(processOne, [s.ledger, s.extension.tokenTotalSupply], lst);
+   let s = FA2.Storage.set_ledger(s)(ledger);
+   [Constants.noOperation,{ ...s, extension : Extension.setSupply(s.extension, supply) }]
+};
+
+// TZIP-17
+let permit = ([permits, s] : [list<permitParams>, storage]) : result => {
+    let processPermit = ([ext, permit] : [Extension.t, permitParams]) : Extension.t => {
+        let [pub_key, [sig, hash_]] = permit;
+        let packed = Bytes.pack ([(Tezos.get_chain_id()), Tezos.get_self_address()], [ext.counter, hash_]);
+        if (Crypto.check(pub_key, sig, packed)) {
+            let sender_ = Tezos.address (Tezos.implicit_account (Crypto.hash_key(pub_key)));
+            let paramPermitKey = [sender_, hash_];
+            let extModified = match (Big_map.find_opt(paramPermitKey, ext.permits), {
+                None: () => {
+                    Extension.addPermit(ext, paramPermitKey)
+                },
+                Some: (submission_timestamp: timestamp) => {
+                    Extension._checkNotExpired(s.extension, submission_timestamp, paramPermitKey);
+                    Extension.updatePermit(ext, paramPermitKey)
+                }
+            });
+            return extModified;
+        } else {
+            return (Michelson `{FAILWITH}` as ((n: [string, bytes]) => Extension.t)) ([Errors.missigned, packed])
+        }
+    };
+    let extension = List.fold_left(processPermit, s.extension, permits);
+    [Constants.noOperation, { ...s, extension : extension }]
+};
+
+// TZIP-17
+let setExpiry = ([p, s] : [expiryParams, storage]) : result => {
+    let [userAddress, [seconds, permitHashOpt]] = p;
+    let _checkMaxExpiry : unit = assert_with_error((seconds <= s.extension.maxExpiry), Errors.maxSecondsExceeded);
+    let _checkSender : unit = assert_with_error((Tezos.get_sender() == userAddress), Errors.forbiddenExpiryUpdate);
+    match (permitHashOpt, {
+        None: () => {
+            return [
+                Constants.noOperation,
+                {...s, extension : { ...s.extension, 
+                    userExpiries : Big_map.add(userAddress, (Some(seconds)), s.extension.userExpiries) }}
+            ];                        
+        },
+        Some: (permit_hash: bytes) => {
+            return [
+                Constants.noOperation,
+                { ...s, extension : {...s.extension, permitExpiries : Big_map.add(
+                    [userAddress, permit_hash],
+                    (Some(seconds)),
+                    s.extension.permitExpiries)
+                }}
+            ]
+        }
+    });
+};
+
+// TZIP-17 implementation of TZIP-12 Transfer
+let transferPermitted = ([t, s]: [FA2.transfer, storage]) : result => {
```

Comment:
If we break the dependency between `extensions` and FA2 the permission and the transfer should be composed providing a `transferPermitted` behaviour. This should also prevent code duplication.

Reply:
agreed, `transferPermitted` dependency to FA2 has been removed